README.md
# redislock-rs - Distributed locks with Redis

![GitHub Workflow Status](https://img.shields.io/github/workflow/status/hxuchen/redislock/CI)
![Crates.io](https://img.shields.io/crates/v/redislock)

This is an implementation of Redislock, the [distributed locking mechanism][distlock] built on top of Redis. It is more or
less a port of the [Ruby version][redlock.rb].

It includes a sample application in [main.rs](src/main.rs).

## Build

```
cargo build --release
```

## Usage

```rust
use redislock::{random_char, RedisLock};

fn main() {
    let rl = RedisLock::new(vec![
        "redis://127.0.0.1:6380/",
        "redis://127.0.0.1:6381/",
        "redis://127.0.0.1:6382/"]);

    let lock;
    loop {
        let val = random_char(Some(20));
        match rl.lock("mutex".as_bytes(), val, 1000, None, None) {
            Some(l) => {
                lock = l;
                break;
            }
            None => ()
        }
    }

    // Critical section

    rl.unlock(&lock);
}
```

## Tests

Run tests with:

```
cargo test
```

Run sample application with:

```
cargo run --release
```

## Contribute

If you find bugs or want to help otherwise, please [open an issue](https://github.com/hxuchen/redislock/issues).

## Maintainer

* Forked from [badboy/redlock-rs](https://github.com/badboy/redlock-rs), maintained by [@IronC](https://github.com/hxuchen)

## License

BSD. See [LICENSE](LICENSE).

[distlock]: http://redis.io/topics/distlock

[redlock.rb]: https://github.com/antirez/redlock-rb


CHANGELOG.md
### 1.3.0 - 06. June 2022

* Rename to `redislock`
* Value can be manually given now
* Fix some bugs
* Retry count and Retry delay can be manually set

### 1.0.0 - 15. January 2016

* Bump to stable 1.0 release
* Test on all platforms all the time
* Remove use of deprecated function `sleep_ms`

### 0.3.0 - 09. April 2015

* Rename to `redlock`
* Make it work on Rust 1.0 beta

### 0.2.0 - 22. November 2014

* Initial release


tests/tests.rs
extern crate core;

use anyhow::Result;
use redislock::{random_char, Lock, RedisLock};
use std::env;

// static DOCKER: Lazy<Cli> = Lazy::new(Cli::default);
// static CONTAINERS: Lazy<Vec<Container<Redis>>> = Lazy::new(|| {
//     (0..3)
//         .map(|_| DOCKER.run(Redis::default().with_tag("6-alpine")))
//         .collect()
// });
// static ADDRESSES: Lazy<Vec<String>> = Lazy::new(|| match std::env::var("ADDRESSES") {
//     Ok(addresses) => addresses.split(',').map(String::from).collect(),
//     Err(_) => CONTAINERS
//         .iter()
//         .map(|c| format!("redis://localhost:{}", c.get_host_port_ipv4(6379)))
//         .collect(),
// });

fn redis_address() -> Vec<String> {
    match env::var("ADDRESSES") {
        Ok(addresses) => addresses.split(',').map(String::from).collect(),
        Err(_) => panic!("redis ADDRESSES must be exported"),
    }
}

#[test]
fn test_redlock_get_unique_id() -> Result<()> {
    let rl = RedisLock::new(Vec::<String>::new());
    assert_eq!(rl.get_unique_lock_id()?.len(), 20);
    Ok(())
}

#[test]
fn test_redlock_get_unique_id_uniqueness() -> Result<()> {
    let rl = RedisLock::new(Vec::<String>::new());

    let id1 = rl.get_unique_lock_id()?;
    let id2 = rl.get_unique_lock_id()?;

    assert_eq!(20, id1.len());
    assert_eq!(20, id2.len());
    assert_ne!(id1, id2);
    Ok(())
}

#[test]
fn test_redlock_valid_instance() {
    println!("{}", redis_address().join(","));
    let rl = RedisLock::new(redis_address().clone());
    assert_eq!(3, rl.servers.len());
    assert_eq!(2, rl.quorum());
}

#[test]
fn test_redlock_direct_unlock_fails() -> Result<()> {
    println!("{}", redis_address().join(","));
    let rl = RedisLock::new(redis_address().clone());
    let key = rl.get_unique_lock_id()?;

    let val = rl.get_unique_lock_id()?;
    assert!(!rl.unlock_instance(&rl.servers[0], &key, &val));
    Ok(())
}

#[test]
fn test_redlock_direct_unlock_succeeds() -> Result<()> {
    println!("{}", redis_address().join(","));
    let rl = RedisLock::new(redis_address().clone());
    let key = rl.get_unique_lock_id()?;

    let val = rl.get_unique_lock_id()?;
    let mut con = rl.servers[0].get_connection()?;
    redis::cmd("SET").arg(&*key).arg(&*val).execute(&mut con);

    assert!(rl.unlock_instance(&rl.servers[0], &key, &val));
    Ok(())
}

#[test]
fn test_redlock_direct_lock_succeeds() -> Result<()> {
    println!("{}", redis_address().join(","));
    let rl = RedisLock::new(redis_address().clone());
    let key = rl.get_unique_lock_id()?;

    let val = random_char(Some(20));
    let mut con = rl.servers[0].get_connection()?;

    redis::cmd("DEL").arg(&*key).execute(&mut con);
    assert!(rl.lock_instance(&rl.servers[0], &*key, &*val, 1000));
    Ok(())
}

#[test]
fn test_redlock_unlock() -> Result<()> {
    println!("{}", redis_address().join(","));
    let rl = RedisLock::new(redis_address().clone());
    let key = rl.get_unique_lock_id()?;

    let val = rl.get_unique_lock_id()?;
    let mut con = rl.servers[0].get_connection()?;
    let _: () = redis::cmd("SET")
        .arg(&*key)
        .arg(&*val)
        .query(&mut con)
        .unwrap();

    let lock = Lock {
        lock_manager: &rl,
        resource: key,
        val,
        validity_time: 0,
    };
    rl.unlock(&lock);
    Ok(())
}

#[test]
fn test_redlock_lock() -> Result<()> {
    println!("{}", redis_address().join(","));
    let rl = RedisLock::new(redis_address().clone());

    let key = rl.get_unique_lock_id()?;
    let val = random_char(Some(20));

    match rl.lock(&key, val, 1000, None, None) {
        Some(lock) => {
            assert_eq!(key, lock.resource);
            assert_eq!(20, lock.val.len());
            assert!(lock.validity_time > 900);
            assert!(
                lock.validity_time > 900,
                "validity time: {}",
                lock.validity_time
            );
        }
        None => panic!("Lock failed"),
    };
    Ok(())
}

#[test]
fn test_redlock_lock_unlock() -> Result<()> {
    println!("{}", redis_address().join(","));
    let rl = RedisLock::new(redis_address().clone());
    let rl2 = RedisLock::new(redis_address().clone());

    let key = rl.get_unique_lock_id()?;
    let val1 = random_char(Some(20));
    let val2 = random_char(Some(20));

    let lock = rl.lock(&key, val1.clone(), 1000, None, None).unwrap();
    assert!(
        lock.validity_time > 900,
        "validity time: {}",
        lock.validity_time
    );

    if let Some(_l) = rl2.lock(&key, val2.clone(), 1000, None, None) {
        panic!("Lock acquired, even though it should be locked")
    }

    rl.unlock(&lock);

    match rl2.lock(&key, val2, 1000, None, None) {
        Some(l) => assert!(l.validity_time > 900),
        None => panic!("Lock couldn't be acquired"),
    }
    Ok(())
}

#[test]
fn test_redlock_lock_unlock_raii() -> Result<()> {
    println!("{}", redis_address().join(","));
    let rl = RedisLock::new(redis_address().clone());
    let rl2 = RedisLock::new(redis_address().clone());

    let key = rl.get_unique_lock_id()?;
    let val1 = random_char(Some(20));
    let val2 = random_char(Some(20));
    {
        let lock_guard = rl.acquire(&key, val1.clone(), 1000, None, None);
        let lock = &lock_guard.lock;
        assert!(
            lock.validity_time > 900,
            "validity time: {}",
            lock.validity_time
        );

        if let Some(_l) = rl2.lock(&key, val2.clone(), 1000, None, None) {
            panic!("Lock acquired, even though it should be locked")
        }
    }

    match rl2.lock(&key, val2.clone(), 1000, None, None) {
        Some(l) => assert!(l.validity_time > 900),
        None => panic!("Lock couldn't be acquired"),
    }
    Ok(())
}


src/lib.rs
//! redislock is an implementation of the [distributed locking
//! mechanism](http://redis.io/topics/distlock) built on top of Redis.
//!
//! It is more or less a port of the [Ruby version](https://github.com/antirez/redlock-rb).
//!
//! # Basic Operation
//! ```rust,no_run
//! # use redislock::{random_char, RedisLock};
//! let rl = RedisLock::new(vec![
//!     "redis://127.0.0.1:6380/",
//!     "redis://127.0.0.1:6381/",
//!     "redis://127.0.0.1:6382/"]);
//!
//! let lock;
//! loop {
//!   let val = random_char(Some(20));
//!   match rl.lock("mutex".as_bytes(), val, 1000, None, None) {
//!     Some(l) => { lock = l; break; }
//!     None => ()
//!   }
//! }
//!
//! // Critical section
//!
//! rl.unlock(&lock);
//! ```

mod redislock;

pub use crate::redislock::{Lock, RedisLock};
use rand::Rng;

const CHARSET: &[u8] = b"ABCDEFGHIJKLMNOPQRSTUVWXYZ\
                            abcdefghijklmnopqrstuvwxyz\
                            0123456789)(*&^%$#@!~";

const DEFAULT_RANDOM_LEN: usize = 20;

pub fn random_char(len: Option<usize>) -> Vec<u8> {
    let mut rng = rand::thread_rng();
    let length: usize;
    if let Some(len) = len {
        if len > 0 {
            length = len
        } else {
            length = DEFAULT_RANDOM_LEN
        }
    } else {
        length = DEFAULT_RANDOM_LEN
    }

    let words: Vec<u8> = (0..length)
        .map(|_| {
            let idx = rng.gen_range(0..CHARSET.len());
            CHARSET[idx] as u8
        })
        .collect();
    words
}

#[cfg(test)]
mod test {
    use crate::random_char;
    use anyhow::Result;

    #[test]
    fn test_random_char() -> Result<()> {
        let words1 = random_char(Some(20));
        println!("{:?}", words1);
        let words2 = random_char(Some(20));
        println!("{:?}", words2);
        assert_eq!(20, words1.len());
        assert_eq!(20, words2.len());
        assert_ne!(words1, words2);
        println!("{}", String::from_utf8(words1).unwrap());
        println!("{}", String::from_utf8(words2).unwrap());
        Ok(())
    }
}


src/redislock.rs
use std::fs::File;
use std::io::{self, Read};
use std::thread::sleep;
use std::time::{Duration, Instant};

use rand::{thread_rng, Rng};
use redis::Value::Okay;
use redis::{Client, IntoConnectionInfo, RedisResult, Value};

const DEFAULT_RETRY_COUNT: u32 = 3;
const DEFAULT_RETRY_DELAY: u32 = 200;
const CLOCK_DRIFT_FACTOR: f32 = 0.01;
const UNLOCK_SCRIPT: &str = r"if redis.call('get',KEYS[1]) == ARGV[1] then
                                return redis.call('del',KEYS[1])
                              else
                                return 0
                              end";

/// The lock manager.
///
/// Implements the necessary functionality to acquire and release locks
/// and handles the Redis connections.
#[derive(Debug, Clone)]
pub struct RedisLock {
    /// List of all Redis clients
    pub servers: Vec<Client>,
    quorum: u32,
    retry_count: u32,
    retry_delay: u32,
}

pub struct Lock<'a> {
    /// The resource to lock. Will be used as the key in Redis.
    pub resource: Vec<u8>,
    /// The value for this lock.
    pub val: Vec<u8>,
    /// Time the lock is still valid.
    /// Should only be slightly smaller than the requested TTL.
    pub validity_time: usize,
    /// Used to limit the lifetime of a lock to its lock manager.
    pub lock_manager: &'a RedisLock,
}

pub struct RedisLockGuard<'a> {
    pub lock: Lock<'a>,
}

impl Drop for RedisLockGuard<'_> {
    fn drop(&mut self) {
        self.lock.lock_manager.unlock(&self.lock);
    }
}

impl RedisLock {
    /// Create a new lock manager instance, defined by the given Redis connection uris.
    /// Quorum is defined to be N/2+1, with N being the number of given Redis instances.
    ///
    /// Sample URI: `"redis://127.0.0.1:6379"`
    pub fn new<T: AsRef<str> + IntoConnectionInfo>(uris: Vec<T>) -> RedisLock {
        let quorum = (uris.len() as u32) / 2 + 1;

        let servers: Vec<Client> = uris
            .into_iter()
            .map(|uri| Client::open(uri).unwrap())
            .collect();

        RedisLock {
            servers,
            quorum,
            retry_count: DEFAULT_RETRY_COUNT,
            retry_delay: DEFAULT_RETRY_DELAY,
        }
    }

    /// get quorum
    pub fn quorum(&self) -> u32 {
        self.quorum
    }

    /// Get 20 random bytes from `/dev/urandom`.
    pub fn get_unique_lock_id(&self) -> io::Result<Vec<u8>> {
        let file = File::open("/dev/urandom")?;
        let mut buf = Vec::with_capacity(20);
        match file.take(20).read_to_end(&mut buf) {
            Ok(20) => Ok(buf),
            Ok(_) => Err(io::Error::new(
                io::ErrorKind::Other,
                "Can't read enough random bytes",
            )),
            Err(e) => Err(e),
        }
    }

    /// Set retry count and retry delay.
    ///
    /// Retry count defaults to `3`.
    /// Retry delay defaults to `200`.
    pub fn set_retry(&mut self, count: u32, delay: u32) {
        self.retry_count = count;
        self.retry_delay = delay;
    }

    pub fn lock_instance(
        &self,
        client: &redis::Client,
        resource: &[u8],
        val: &[u8],
        ttl: usize,
    ) -> bool {
        let mut con = match client.get_connection() {
            Err(_) => return false,
            Ok(val) => val,
        };
        let result: RedisResult<Value> = redis::cmd("SET")
            .arg(resource)
            .arg(val)
            .arg("nx")
            .arg("px")
            .arg(ttl)
            .query(&mut con);
        match result {
            Ok(Okay) => true,
            Ok(_) | Err(_) => false,
        }
    }

    /// Acquire the lock for the given resource and the requested TTL.
    ///
    /// If it succeeds, a `Lock` instance is returned,
    /// including the value and the validity time
    ///
    /// If it fails. `None` is returned.
    /// A user should retry after a short wait time.
    pub fn lock(
        &self,
        resource: &[u8],
        val: Vec<u8>,
        ttl: usize,
        retry_count: Option<u32>,
        retry_delay: Option<u32>,
    ) -> Option<Lock> {
        // let val = self.get_unique_lock_id().unwrap();

        let retry_count = {
            || {
                if let Some(count) = retry_count {
                    if count > 0 {
                        count
                    } else {
                        self.retry_count
                    }
                } else {
                    self.retry_count
                }
            }
        }();

        for _ in 0..retry_count {
            let mut n = 0;
            let start_time = Instant::now();
            for client in &self.servers {
                if self.lock_instance(client, resource, &val, ttl) {
                    n += 1;
                }
            }

            let drift = (ttl as f32 * CLOCK_DRIFT_FACTOR) as usize + 2;
            let elapsed = start_time.elapsed();
            let validity_time = ttl
                .saturating_sub(drift)
                .saturating_sub(elapsed.as_secs() as usize * 1000)
                .saturating_sub(elapsed.subsec_nanos() as usize / 1_000_000);

            if n >= self.quorum && validity_time > 0 {
                return Some(Lock {
                    lock_manager: self,
                    resource: resource.to_vec(),
                    val,
                    validity_time,
                });
            } else {
                for client in &self.servers {
                    self.unlock_instance(client, resource, &val);
                }
            }

            if let Some(retry_delay) = retry_delay {
                sleep(Duration::from_millis(u64::from(retry_delay)))
            } else {
                let mut rng = thread_rng();
                let n = rng.gen_range(0..self.retry_delay);
                sleep(Duration::from_millis(u64::from(n)));
            }
        }
        None
    }

    /// Acquire the lock for the given resource and the requested TTL. \
    /// Will wait and yield current task (tokio runtime) until the lock \
    /// is acquired
    ///
    /// Returns a `RedisLockGuard` instance which is a RAII wrapper for \
    /// the old `Lock` object
    #[cfg(feature = "async")]
    pub async fn acquire_async(
        &self,
        resource: &[u8],
        val: Vec<u8>,
        ttl: usize,
        retry_count: Option<u32>,
        retry_delay: Option<u32>,
    ) -> RedisLockGuard<'_> {
        let lock;
        loop {
            match self.lock(resource, val.clone(), ttl, retry_count, retry_delay) {
                Some(l) => {
                    lock = l;
                    break;
                }
                None => tokio::task::yield_now().await,
            }
        }
        RedisLockGuard { lock }
    }

    pub fn acquire(
        &self,
        resource: &[u8],
        val: Vec<u8>,
        ttl: usize,
        retry_count: Option<u32>,
        retry_delay: Option<u32>,
    ) -> RedisLockGuard<'_> {
        let lock;
        loop {
            if let Some(l) = self.lock(resource, val.clone(), ttl, retry_count, retry_delay) {
                lock = l;
                break;
            }
        }
        RedisLockGuard { lock }
    }

    pub fn unlock_instance(&self, client: &redis::Client, resource: &[u8], val: &[u8]) -> bool {
        let mut con = match client.get_connection() {
            Err(_) => return false,
            Ok(val) => val,
        };
        let script = redis::Script::new(UNLOCK_SCRIPT);
        let result: RedisResult<i32> = script.key(resource).arg(val).invoke(&mut con);
        match result {
            Ok(val) => val == 1,
            Err(_) => false,
        }
    }

    /// Unlock the given lock.
    ///
    /// Unlock is best effort. It will simply try to contact all instances
    /// and remove the key.
    pub fn unlock(&self, lock: &Lock) {
        for client in &self.servers {
            self.unlock_instance(client, &lock.resource, &lock.val);
        }
    }
}


examples/basic.rs
use rand::Rng;
use redislock::{random_char, RedisLock};

use std::env;
use std::sync::mpsc::channel;
use std::thread;
use std::time::Duration;

pub fn main() {
    let mut number_of_workers = 10;
    let mut number_of_incrs = 55;
    let resource = "incr_mutex";
    let incr_key = "incr_key";

    let mut args = env::args();
    if args.len() == 3 {
        number_of_workers = args.next().unwrap().parse().unwrap();
        number_of_incrs = args.next().unwrap().parse().unwrap();
    }

    let result: u32 = number_of_workers * number_of_incrs;

    let (tx, rx) = channel();

    println!(
        "Starting {} workers, each incrementing {} times.",
        number_of_workers, number_of_incrs
    );

    let client = redis::Client::open("redis://127.0.0.1:6380/").unwrap();
    let mut con = client.get_connection().unwrap();
    redis::cmd("DEL").arg(incr_key).execute(&mut con);

    for _ in 0..number_of_workers {
        let tx = tx.clone();
        let _ = thread::spawn(move || {
            let mut rng = rand::thread_rng();

            let rl = RedisLock::new(vec![
                "redis://127.0.0.1:6380/",
                "redis://127.0.0.1:6381/",
                "redis://127.0.0.1:6382/",
            ]);
            let mut con = rl.servers[0].get_connection().unwrap();

            for _ in 0..number_of_incrs {
                let lock;
                loop {
                    let rand_val = random_char(Some(20));
                    if let Some(l) = rl.lock(resource.as_bytes(), rand_val, 1000, None, None) {
                        lock = l;
                        break;
                    }
                }
                let val: i32 = redis::cmd("GET").arg(incr_key).query(&mut con).unwrap_or(0);

                let n = rng.gen_range(0..5);
                thread::sleep(Duration::from_millis(n));

                redis::cmd("SET")
                    .arg(incr_key)
                    .arg(val + 1)
                    .execute(&mut con);

                rl.unlock(&lock);
            }

            tx.send(()).unwrap();
        });
    }

    let mut i = 0;
    loop {
        if i == number_of_workers {
            break;
        }

        let _ = rx.recv();
        i += 1;
    }

    let actual_result: u32 = redis::cmd("GET").arg(incr_key).query(&mut con).unwrap_or(0);

    println!("Expected result: {}", result);
    println!("Actual result:   {}", actual_result);
    if result == actual_result {
        println!("Everything fine! \\o/");
    } else {
        println!("Something is broken. /o\\");
    }
}


src/redlock.rs
use std::fs::File;
use std::io::{self, Read};
use std::thread::sleep;
use std::time::{Duration, Instant};

use rand::{thread_rng, Rng};
use redis::Value::Okay;
use redis::{Client, IntoConnectionInfo, RedisResult, Value};

const DEFAULT_RETRY_COUNT: u32 = 3;
const DEFAULT_RETRY_DELAY: u32 = 200;
const CLOCK_DRIFT_FACTOR: f32 = 0.01;
const UNLOCK_SCRIPT: &str = r"if redis.call('get',KEYS[1]) == ARGV[1] then
                                return redis.call('del',KEYS[1])
                              else
                                return 0
                              end";

/// The lock manager.
///
/// Implements the necessary functionality to acquire and release locks
/// and handles the Redis connections.
#[derive(Debug, Clone)]
pub struct RedLock {
    /// List of all Redis clients
    pub servers: Vec<Client>,
    quorum: u32,
    retry_count: u32,
    retry_delay: u32,
}

pub struct Lock<'a> {
    /// The resource to lock. Will be used as the key in Redis.
    pub resource: Vec<u8>,
    /// The value for this lock.
    pub val: Vec<u8>,
    /// Time the lock is still valid.
    /// Should only be slightly smaller than the requested TTL.
    pub validity_time: usize,
    /// Used to limit the lifetime of a lock to its lock manager.
    pub lock_manager: &'a RedLock,
}

pub struct RedLockGuard<'a> {
    pub lock: Lock<'a>,
}

impl Drop for RedLockGuard<'_> {
    fn drop(&mut self) {
        self.lock.lock_manager.unlock(&self.lock);
    }
}

impl RedLock {
    /// Create a new lock manager instance, defined by the given Redis connection uris.
    /// Quorum is defined to be N/2+1, with N being the number of given Redis instances.
    ///
    /// Sample URI: `"redis://127.0.0.1:6379"`
    pub fn new<T: AsRef<str> + IntoConnectionInfo>(uris: Vec<T>) -> RedLock {
        let quorum = (uris.len() as u32) / 2 + 1;

        let servers: Vec<Client> = uris
            .into_iter()
            .map(|uri| Client::open(uri).unwrap())
            .collect();

        RedLock {
            servers,
            quorum,
            retry_count: DEFAULT_RETRY_COUNT,
            retry_delay: DEFAULT_RETRY_DELAY,
        }
    }

    /// Get 20 random bytes from `/dev/urandom`.
    pub fn get_unique_lock_id(&self) -> io::Result<Vec<u8>> {
        let file = File::open("/dev/urandom")?;
        let mut buf = Vec::with_capacity(20);
        match file.take(20).read_to_end(&mut buf) {
            Ok(20) => Ok(buf),
            Ok(_) => Err(io::Error::new(
                io::ErrorKind::Other,
                "Can't read enough random bytes",
            )),
            Err(e) => Err(e),
        }
    }

    /// Set retry count and retry delay.
    ///
    /// Retry count defaults to `3`.
    /// Retry delay defaults to `200`.
    pub fn set_retry(&mut self, count: u32, delay: u32) {
        self.retry_count = count;
        self.retry_delay = delay;
    }

    fn lock_instance(
        &self,
        client: &redis::Client,
        resource: &[u8],
        val: &[u8],
        ttl: usize,
    ) -> bool {
        let mut con = match client.get_connection() {
            Err(_) => return false,
            Ok(val) => val,
        };
        let result: RedisResult<Value> = redis::cmd("SET")
            .arg(resource)
            .arg(val)
            .arg("nx")
            .arg("px")
            .arg(ttl)
            .query(&mut con);
        match result {
            Ok(Okay) => true,
            Ok(_) | Err(_) => false,
        }
    }

    /// Acquire the lock for the given resource and the requested TTL.
    ///
    /// If it succeeds, a `Lock` instance is returned,
    /// including the value and the validity time
    ///
    /// If it fails. `None` is returned.
    /// A user should retry after a short wait time.
    pub fn lock(
        &self,
        resource: &[u8],
        val: Vec<u8>,
        ttl: usize,
        retry_count: Option<u32>,
        retry_delay: Option<u32>,
    ) -> Option<Lock> {
        // let val = self.get_unique_lock_id().unwrap();

        let retry_count = {
            || {
                if let Some(count) = retry_count {
                    if count > 0 {
                        count
                    } else {
                        self.retry_count
                    }
                } else {
                    self.retry_count
                }
            }
        }();

        for _ in 0..retry_count {
            let mut n = 0;
            let start_time = Instant::now();
            for client in &self.servers {
                if self.lock_instance(client, resource, &val, ttl) {
                    n += 1;
                }
            }

            let drift = (ttl as f32 * CLOCK_DRIFT_FACTOR) as usize + 2;
            let elapsed = start_time.elapsed();
            let validity_time = ttl
                .saturating_sub(drift)
                .saturating_sub(elapsed.as_secs() as usize * 1000)
                .saturating_sub(elapsed.subsec_nanos() as usize / 1_000_000);

            if n >= self.quorum && validity_time > 0 {
                return Some(Lock {
                    lock_manager: self,
                    resource: resource.to_vec(),
                    val,
                    validity_time,
                });
            } else {
                for client in &self.servers {
                    self.unlock_instance(client, resource, &val);
                }
            }

            if let Some(retry_delay) = retry_delay {
                sleep(Duration::from_millis(u64::from(retry_delay)))
            } else {
                let mut rng = thread_rng();
                let n = rng.gen_range(0..self.retry_delay);
                sleep(Duration::from_millis(u64::from(n)));
            }
        }
        None
    }

    /// Acquire the lock for the given resource and the requested TTL. \
    /// Will wait and yield current task (tokio runtime) until the lock \
    /// is acquired
    ///
    /// Returns a `RedLockGuard` instance which is a RAII wrapper for \
    /// the old `Lock` object
    #[cfg(feature = "async")]
    pub async fn acquire_async(
        &self,
        resource: &[u8],
        val: Vec<u8>,
        ttl: usize,
        retry_count: Option<u32>,
        retry_delay: Option<u32>,
    ) -> RedLockGuard<'_> {
        let lock;
        loop {
            match self.lock(resource, val.clone(), ttl, retry_count, retry_delay) {
                Some(l) => {
                    lock = l;
                    break;
                }
                None => tokio::task::yield_now().await,
            }
        }
        RedLockGuard { lock }
    }

    pub fn acquire(
        &self,
        resource: &[u8],
        val: Vec<u8>,
        ttl: usize,
        retry_count: Option<u32>,
        retry_delay: Option<u32>,
    ) -> RedLockGuard<'_> {
        let lock;
        loop {
            if let Some(l) = self.lock(resource, val.clone(), ttl, retry_count, retry_delay) {
                lock = l;
                break;
            }
        }
        RedLockGuard { lock }
    }

    fn unlock_instance(&self, client: &redis::Client, resource: &[u8], val: &[u8]) -> bool {
        let mut con = match client.get_connection() {
            Err(_) => return false,
            Ok(val) => val,
        };
        let script = redis::Script::new(UNLOCK_SCRIPT);
        let result: RedisResult<i32> = script.key(resource).arg(val).invoke(&mut con);
        match result {
            Ok(val) => val == 1,
            Err(_) => false,
        }
    }

    /// Unlock the given lock.
    ///
    /// Unlock is best effort. It will simply try to contact all instances
    /// and remove the key.
    pub fn unlock(&self, lock: &Lock) {
        for client in &self.servers {
            self.unlock_instance(client, &lock.resource, &lock.val);
        }
    }
}

#[cfg(test)]
mod tests {
    use crate::random_char;
    use anyhow::Result;
    use once_cell::sync::Lazy;
    use testcontainers::clients::Cli;
    use testcontainers::images::redis::Redis;
    use testcontainers::{Container, Docker};

    use super::*;

    static DOCKER: Lazy<Cli> = Lazy::new(Cli::default);
    static CONTAINERS: Lazy<Vec<Container<Cli, Redis>>> = Lazy::new(|| {
        (0..3)
            .map(|_| DOCKER.run(Redis::default().with_tag("6-alpine")))
            .collect()
    });
    static ADDRESSES: Lazy<Vec<String>> = Lazy::new(|| match std::env::var("ADDRESSES") {
        Ok(addresses) => addresses.split(',').map(String::from).collect(),
        Err(_) => CONTAINERS
            .iter()
            .map(|c| format!("redis://localhost:{}", c.get_host_port(6379).unwrap()))
            .collect(),
    });

    #[test]
    fn test_redlock_get_unique_id() -> Result<()> {
        let rl = RedLock::new(Vec::<String>::new());
        assert_eq!(rl.get_unique_lock_id()?.len(), 20);
        Ok(())
    }

    #[test]
    fn test_redlock_get_unique_id_uniqueness() -> Result<()> {
        let rl = RedLock::new(Vec::<String>::new());

        let id1 = rl.get_unique_lock_id()?;
        let id2 = rl.get_unique_lock_id()?;

        assert_eq!(20, id1.len());
        assert_eq!(20, id2.len());
        assert_ne!(id1, id2);
        Ok(())
    }

    #[test]
    fn test_redlock_valid_instance() {
        println!("{}", ADDRESSES.join(","));
        let rl = RedLock::new(ADDRESSES.clone());
        assert_eq!(3, rl.servers.len());
        assert_eq!(2, rl.quorum);
    }

    #[test]
    fn test_redlock_direct_unlock_fails() -> Result<()> {
        println!("{}", ADDRESSES.join(","));
        let rl = RedLock::new(ADDRESSES.clone());
        let key = rl.get_unique_lock_id()?;

        let val = rl.get_unique_lock_id()?;
        assert!(!rl.unlock_instance(&rl.servers[0], &key, &val));
        Ok(())
    }

    #[test]
    fn test_redlock_direct_unlock_succeeds() -> Result<()> {
        println!("{}", ADDRESSES.join(","));
        let rl = RedLock::new(ADDRESSES.clone());
        let key = rl.get_unique_lock_id()?;

        let val = rl.get_unique_lock_id()?;
        let mut con = rl.servers[0].get_connection()?;
        redis::cmd("SET").arg(&*key).arg(&*val).execute(&mut con);

        assert!(rl.unlock_instance(&rl.servers[0], &key, &val));
        Ok(())
    }

    #[test]
    fn test_redlock_direct_lock_succeeds() -> Result<()> {
        println!("{}", ADDRESSES.join(","));
        let rl = RedLock::new(ADDRESSES.clone());
        let key = rl.get_unique_lock_id()?;

        let val = random_char(Some(20));
        let mut con = rl.servers[0].get_connection()?;

        redis::cmd("DEL").arg(&*key).execute(&mut con);
        assert!(rl.lock_instance(&rl.servers[0], &*key, &*val, 1000));
        Ok(())
    }

    #[test]
    fn test_redlock_unlock() -> Result<()> {
        println!("{}", ADDRESSES.join(","));
        let rl = RedLock::new(ADDRESSES.clone());
        let key = rl.get_unique_lock_id()?;

        let val = rl.get_unique_lock_id()?;
        let mut con = rl.servers[0].get_connection()?;
        let _: () = redis::cmd("SET")
            .arg(&*key)
            .arg(&*val)
            .query(&mut con)
            .unwrap();

        let lock = Lock {
            lock_manager: &rl,
            resource: key,
            val,
            validity_time: 0,
        };
        rl.unlock(&lock);
        Ok(())
    }

    #[test]
    fn test_redlock_lock() -> Result<()> {
        println!("{}", ADDRESSES.join(","));
        let rl = RedLock::new(ADDRESSES.clone());

        let key = rl.get_unique_lock_id()?;
        let val = random_char(Some(20));

        match rl.lock(&key, val, 1000, None, None) {
            Some(lock) => {
                assert_eq!(key, lock.resource);
                assert_eq!(20, lock.val.len());
                assert!(lock.validity_time > 900);
                assert!(
                    lock.validity_time > 900,
                    "validity time: {}",
                    lock.validity_time
                );
            }
            None => panic!("Lock failed"),
        };
        Ok(())
    }

    #[test]
    fn test_redlock_lock_unlock() -> Result<()> {
        println!("{}", ADDRESSES.join(","));
        let rl = RedLock::new(ADDRESSES.clone());
        let rl2 = RedLock::new(ADDRESSES.clone());

        let key = rl.get_unique_lock_id()?;
        let val1 = random_char(Some(20));
        let val2 = random_char(Some(20));

        let lock = rl.lock(&key, val1.clone(), 1000, None, None).unwrap();
        assert!(
            lock.validity_time > 900,
            "validity time: {}",
            lock.validity_time
        );

        if let Some(_l) = rl2.lock(&key, val2.clone(), 1000, None, None) {
            panic!("Lock acquired, even though it should be locked")
        }

        rl.unlock(&lock);

        match rl2.lock(&key, val2, 1000, None, None) {
            Some(l) => assert!(l.validity_time > 900),
            None => panic!("Lock couldn't be acquired"),
        }
        Ok(())
    }

    #[test]
    fn test_redlock_lock_unlock_raii() -> Result<()> {
        println!("{}", ADDRESSES.join(","));
        let rl = RedLock::new(ADDRESSES.clone());
        let rl2 = RedLock::new(ADDRESSES.clone());

        let key = rl.get_unique_lock_id()?;
        let val1 = random_char(Some(20));
        let val2 = random_char(Some(20));
        {
            let lock_guard = rl.acquire(&key, val1.clone(), 1000, None, None);
            let lock = &lock_guard.lock;
            assert!(
                lock.validity_time > 900,
                "validity time: {}",
                lock.validity_time
            );

            if let Some(_l) = rl2.lock(&key, val2.clone(), 1000, None, None) {
                panic!("Lock acquired, even though it should be locked")
            }
        }

        match rl2.lock(&key, val2.clone(), 1000, None, None) {
            Some(l) => assert!(l.validity_time > 900),
            None => panic!("Lock couldn't be acquired"),
        }
        Ok(())
    }
}
